pkg/services/rpcsrv/util.go

```diff
@@ -18,3 +24,46 @@ func checkInt32(i int) error {
 	}
 	return nil
 }
+
+func (c notificationComparatorFilter) EventID() neorpc.EventID {
+	return c.id
+}
+
+func (c notificationComparatorFilter) Filter() neorpc.SubscriptionFilter {
+	return c.filter
+}
+
+func (c notificationEventContainer) EventID() neorpc.EventID {
+	return neorpc.NotificationEventID
+}
+
+func (c notificationEventContainer) EventPayload() any {
+	return c.ntf
+}
+
+func processAppExecResults(aers []state.AppExecResult, filter *neorpc.NotificationFilter) []state.ContainedNotificationEvent {
+	var notifications []state.ContainedNotificationEvent
+	for _, aer := range aers {
+		if aer.VMState == vmstate.Halt {
+			notifications = append(notifications, filterEvents(aer.Events, aer.Container, filter)...)
+		}
+	}
+	return notifications
+}
+
+func filterEvents(events []state.NotificationEvent, container util.Uint256, filter *neorpc.NotificationFilter) []state.ContainedNotificationEvent {
+	var notifications []state.ContainedNotificationEvent
+	for _, evt := range events {
+		ntf := state.ContainedNotificationEvent{
+			Container:         container,
+			NotificationEvent: evt,
+		}
+		if filter == nil || rpcevent.Matches(&notificationComparatorFilter{
```

Comment:
`filter == nil` check is excessive. `rpcevent.Matche` perfectly handle this case.